tests/linearizability/traffic.go

```diff
@@ -87,6 +89,24 @@ func (t readWriteSingleKey) Write(ctx context.Context, c *recordingClient, limit
 			expectValue = string(lastValues[0].Value)
 		}
 		err = c.Txn(putCtx, key, expectValue, newValue)
+	case PutWithLease:
+		leaseId := lm.LeaseId(cid)
+		if leaseId == 0 {
+			leaseId, err = c.LeaseGrant(ctx, t.leaseTTL)
+			lm.AddLeaseId(cid, leaseId)
+		}
+		if leaseId != 0 {
+			err = c.PutWithLease(putCtx, key, newValue, leaseId)
+		}
+	case LeaseRevoke:
+		leaseId := lm.LeaseId(cid)
+		if leaseId != 0 {
+			err = c.LeaseRevoke(putCtx, leaseId)
+			//if LeaseRevoke has failed, do not remove the mapping.
+			if err == nil {
+				lm.RemoveLeaseId(cid)
+			}
+		}
```

Comment:
Should you execute `c.LeaseGrant` as well if there is no a leaseId related to the clientId in this case (`LeaseRevoke`), just in the same way as you process `PutWithLease`? Otherwise the `(readWriteSingleKey) Write` will do nothing in this case.

The [Delete operation](https://github.com/etcd-io/etcd/blob/ff898640a5c9bad0bb99a74d0799f810d54b3586/tests/linearizability/traffic.go#L82) might delete nothing if it happens before the related `PUT` operation. But there is always a communication with the etcdserver. 

I agree that the existing implementation is simpler. 

Anyway, I am  OK to discuss this in a separate PR.